[3.4] Split Vertex Buffer Stream in Positions and Attributes

The-O-King opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

Provide an import option for enabling vertex stream splitting - where the vertex 
buffer is organized into a contiguous block of position data and a contiguous
block of interleaved attribute data

This is an optimization particularly for mobile, tile-based renderers which can
use less memory bandwidth during their binning phase

Implementation of https://github.com/godotengine/godot-proposals/issues/2350

The-O-King committed:
Split Vertex Position and Attribute Streams

Implemented splitting of vertex positions and attributes in the vertex
buffer

Positions are sequential at the start of the buffer, followed by the
additional attributes which are interleaved

Made a project setting which enables/disabled the buffer formatting
throughout the project

Implemented in both GLES2 and GLES3

This improves performance particularly on tile-based GPUs as well as
cache performance for something like shadow mapping which only needs
position data

Updated Docs and Project Setting

fire commented:
Is there a godot master version I can review?

The-O-King commented:
At the moment I only have this implemented in 3.2 since 4.0 is still undergoing work on the rendering side from what I understand

I think discussion is still ongoing as to what/if 4.0 will need something like this implemented (see the proposal linked for discussion)

clayjohn commented:
What do you think about making this a ProjectSetting instead of an import setting. I don't see a reason to import some meshes using the split vertex stream and some without. To me it seems like Desktop should use non-split and mobile should use split. In which case there can be a check in the rasterizer. I think making this a project setting would result in a simpler PR and would be more user-friendly. 

Please let me know if there is a good reason to have this in the importer.

lawnjelly commented:
Also have you got any timing figures for the performance with and without the change on different hardware? It would be kind of nice to have some figures to see the difference.

The-O-King commented:
Ah that's a good thought and not something that I had considered before, looking at it I think that a project setting would make more sense also, I'll look into what that would take! Thank you for pointing me towards that!

For performance, I know that for mobile it helps primarily with memory bandwidth metrics (and was the primary motivation for this change, 50% improvement) rather than gpu frame times, and I did do some cursory testing on desktop (nvidia 2070 super Max-Q laptop) and did not find much in the way of performance difference, but I will try to get more concrete numbers for that as well to share

The-O-King commented:
Implementation question - is there any way to trigger a reimport of all mesh assets when a project setting is changed? At the moment someone could change the project setting, but all previously imported assets would still have the old mesh data format and they would have to reimport the assets manually

The-O-King commented:
> > Implementation question - is there any way to trigger a reimport of all mesh assets when a project setting is changed? At the moment someone could change the project setting, but all previously imported assets would still have the old mesh data format and they would have to reimport the assets manually
> 
> As far as I know, there is no way to do that yet (aside of removing the `.import/` _folder_). This is also an issue when changing the VRAM compressed texture formats in the Project Settings.

Gotcha, thanks for letting me know! Maybe we can make that an option in the future haha :) 

For now I was able to update the implementation to use project settings rather than import settings for splitting the vertex buffer streams, and will have similar functionality to the way that the VRAM texture compression settings will behave then

I also removed the commit that added the mesh compression options since it is much less relevant now to this PR, and created a new PR with that feature addition here https://github.com/godotengine/godot/pull/48625

clayjohn commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -2265,27 +2268,32 @@ void RasterizerStorageGLES2::mesh_add_surface(RID p_mesh, uint32_t p_format, VS:
-					attribs[i].type = _GL_HALF_FLOAT_OES;
-					stride += attribs[i].size * 2;
-					uses_half_float = true;
+					attribs[i].type = GL_HALF_FLOAT;
```
I'm guessing this is a copy-paste error from GLES3. This line is causing the CI to fail because it should be ``_GL_HALF_FLOAT_OES``

clayjohn commented:
Implementation looks good on a preliminary review! 

Have you had any luck getting performance numbers on mobile? I would be really interested to see if there is a measurable difference. The [TPS demo](https://github.com/godotengine/tps-demo) may be a good test project if you don't already have one.

clayjohn commented on servers/visual_server.cpp:
```diff
@@ -329,7 +329,8 @@ RID VisualServer::get_white_texture() {
 
-Error VisualServer::_surface_set_data(Array p_arrays, uint32_t p_format, uint32_t *p_offsets, uint32_t p_stride, PoolVector<uint8_t> &r_vertex_array, int p_vertex_array_len, PoolVector<uint8_t> &r_index_array, int p_index_array_len, AABB &r_aabb, Vector<AABB> &r_bone_aabb) {
+Error VisualServer::_surface_set_data(Array p_arrays, uint32_t p_format, uint32_t *p_offsets, uint32_t *p_stride, PoolVector<uint8_t> &r_vertex_array, int p_vertex_array_len, PoolVector<uint8_t> &r_index_array, int p_index_array_len, AABB &r_aabb, Vector<AABB> &r_bone_aabb) {
+
```
delete this newline. It is failing the style check

clayjohn commented on servers/visual_server.cpp:
```diff
@@ -512,11 +513,12 @@ Error VisualServer::_surface_set_data(Array p_arrays, uint32_t p_format, uint32_
 				} else {
 					for (int i = 0; i < p_vertex_array_len; i++) {
-						memcpy(&vw[p_offsets[ai] + i * p_stride], &src[i], 4 * 4);
+
```
also failing style check. delete this line

clayjohn commented on servers/visual_server.cpp:
```diff
@@ -944,13 +951,24 @@ void VisualServer::mesh_add_surface_from_arrays(RID p_mesh, PrimitiveType p_prim
+					attributes_base_offset = elem_size;
+				}
+
+
```
delete this line

The-O-King commented:
Vertex Stream Splitting Results Mobile - Tested on Pixel 4 - The scene had normal mapped spheres offscreen to isolate the test to just vertex data (so no fragments would be shaded) and was ~68.8million verts

- Vertex Memory Read Bandwidth:
    - Unsplit: 27GB/s 
    - Split: 6.5GB/s
- Vertex Fetch Stalls:
    - Unsplit: 66% 
    - Split 53%
- Average Bytes/Vertex:
    - Unsplit: 48B 
    - Split: 12B
- Frame Time:
    - Unsplit: 120ms
    - Split: 87ms
    - Note that this is a VERY vertex bound scene so this might not be the kind of improvement seen in all games haha

This is a screenshot from AGI with the data - the left side has the unsplit data, the right side has split data
![2021-05-11 (2)](https://user-images.githubusercontent.com/14830461/117857650-fe689680-b25a-11eb-92de-71e6a1f8da9e.png)

I will also try to get data for the TPS demo since that is a more real world test - but these are the theoretical improvements one could see :)

lawnjelly commented:
> You do have to take the buffer format into account if you are updating the raw buffers directly - and yes you would need to make two writes in order to update said data - I myself haven't look at what changes would be needed for that (or the driver implications) but I don't think it necessarily means that we have to rule out dynamic meshes? Unless I'm missing something

I am not convinced there is a compelling reason to apply this to dynamic meshes. Firstly there there seems to be negligible performance increase, so the argument is in terms of power use (which hasn't been measured, this may again be negligible because the pixel shaders are doing the bulk of the work and consuming 99% of the power).

On the other hand splitting the vertices has several potential negative implications with dynamic buffers, which would seem to range from slightly bad to very bad.

e.g.
* Two uploads in the case of partial buffer use. This could be extremely bad for performance (even causing stalls), especially with the patchy driver landscape. Quite a bit of effort has been made in the past year to attempt the opposite of this, in order to solve performance issues. This could also be a repeat of the situation with buffer orphaning. If there are no advantages to this approach, why take the risk of these problems?
* Jumping around in cache to fill these buffers. May or may not be a problem.

clayjohn commented:
@lawnjelly This PR only changes static mesh buffers. Dynamic buffers (e.g. for ImmediateGeometry) are unaffected. Really it just de-interleaves the position attribute of the vertex so that TBDR GPU's can optimize the binning phase (also it opens up one further optimization which is only sending the position section of the buffer for shadow rendering).

The-O-King commented:
I also collected data for desktop and at the moment it looks like performance is basically identical - This was in a static scene with 90,000,000 verts that were offscreen, so I would say having it on desktop will at least keep the same performance

In Nvidia Nsight, the bottleneck here is in VRAM (both have it pinned at 97% SOL) and the L2 cache behavior is also basically the same (46% hit rate in this scene). There was a reduction in frame time (average 43ms to 41ms) but I would chalk that up to occasional power management/throttling

Unsplit:
![2021-05-12 - unsplit - desktop (2)](https://user-images.githubusercontent.com/14830461/118016640-a2b61000-b323-11eb-8730-50e4f6e189c9.png)
Split:
![2021-05-12 - split - desktop (2)](https://user-images.githubusercontent.com/14830461/118016661-a77ac400-b323-11eb-981c-51a4ac822278.png)


I also did a test with the same scene but an additional dynamic light/shadow (but reduced down to 60,000,000 to make it a more reasonable framerate) and I got the same exact results between split and unsplit - although when I was checking things in renderdoc - I did notice that the shadow map generating draw calls had the vertex normal enabled - and I was wondering what they were used for in this scenario? As that could be a good area for improvement on both desktop and mobile

This is also more anecdotal evidence, but I would also note from performance profiling other games on desktop (such as Apex Legends, or the default vertex buffer format in Unreal Engine) they all usually have some sort of vertex buffer stream splitting in place where positions are their own stream and attributes are distributed amongst other separate buffers, grouped by what data is most commonly used together in a shader/renderpass

The-O-King commented:
Ah sorry I think I misunderstood before about the static vs dynamic mesh thing - yes this is applied to static meshes only

I see definitely see where you're coming from in terms of seeing real-world improvements, but I think the TPS scene in particular isn't the best way to show the improvements one would get from something like this because it's unreasonably pixel bound haha, at least on mobile :P I think in all scenarios though it will come out with same or improved performance

We do have the data for static mesh scenes (single or not shouldn't make a difference) on mobile since that was what the first scenario I presented was testing (to be fair it wasn't a "single" static mesh but 30 copies of the same high poly single mesh, which I would argue end up with similar effect)

And for tile based renderers (which are all mobile GPUs) - rendering to a single tile vs offscreen would show the same performance improvements that the first test scenario I presented showed, just with the added pixel shader work for fragments that end up on screen, regardless of the tile - the binning process occurs for ALL vertices before shading fragments (or the rest of the vertex calculations) ever begins

For the power argument, I don't think I have a good way to measure the power improvement without running an example load/app on a device and running it until it dies haha - but we know that's it's a recommendation that both ARM[[1]](https://developer.arm.com/documentation/101897/0200/vertex-shading/attribute-layout)[[2] ](https://developer.arm.com/documentation/dui0555/c/bandwidth-optimizations/about-reducing-bandwidth?lang=en)and [Qualcomm ](https://developer.qualcomm.com/docs/adreno-gpu/developer-guide/gpu/best_practices_vertex.html)make in their best practices guides 
[nvidia generally mentions stream splitting can help](https://developer.nvidia.com/blog/the-peak-performance-analysis-method-for-optimizing-any-gpu-workload/#5.3.4) in some places for doing things like shadow/occlusion passes

lawnjelly commented:
If performance is within a gnat's todger with the split vertex format for static buffers on desktop, then to me that does sound a fair argument for just using split as the path for all static buffers, and not bothering with the split / non-split option. Unless there is a need for backward compatibility. What do you think @clayjohn ? :smile:

clayjohn commented:
@lawnjelly I agree that we need to do more performance analysis before we commit to anything. 

_In theory_ this PR should result in performance benefits in some cases on mobile. And _in theory_ it should degrade performance on desktop. The reduction in performance comes from the cache pressure resulting from reading the vertex position from one end of the vertex array while reading the other vertex attributes from another position in the array. Again, the performance consideration is only theoretical and may pan out to be negligible.

Calinou commented:
@The-O-King Can you upload the vertex-heavy test project you used for testing (not the TPS demo)? This way, we can check the performance impact on desktop platforms.

The-O-King commented:
Oh yes sure thing - here's a google drive link to a zip of the project I've been using to test things - there's a sphere_scene_desktop.tscn which is what I used to measure the performance - currently there's a moving point light with shadows as well for when I was profiling that scenario

https://drive.google.com/file/d/1X_Dc9NJfN6p8xYf4-g64SVuOEJCxJzrW/view?usp=sharing

Let me know if this drive link works, or if there's a better/preferred way to share the project :)

Calinou commented:
> Let me know if this drive link works, or if there's a better/preferred way to share the project :)

The link works, thanks :slightly_smiling_face:

lawnjelly commented:
Nice little discussion of this on twitter today too, and on when the situation changed on this (looks like around 2015/2016?):
https://twitter.com/promit_roy/status/1394847846759911427

lawnjelly commented:
Just been trying the PR on my desktop (Linux Mint 20, Intel HD Graphics 630) and seeing if there's any difference.

With this quick contrived test I can get 129fps without splitting and 126fps with splitting, but I doubt there is much in it in practice. My test was procedurally generating a mesh with 1000000 triangles (3000000 unique verts) with uv, uv2, color, normal and tangent, and very small screen size and identical small triangle size. I don't even know if it uses that many verts or limits to 65535 in fact .. it could do in GLES3 though.

Be interesting to try on more hardware these kinds of tests.
[SplitVertBufferTest.zip](https://github.com/godotengine/godot/files/6513851/SplitVertBufferTest.zip)

The-O-King commented:
I got a chance today to run your test project (really elegant btw I love it haha) and on my laptop which has a RTX 2070 MAX-Q I got identical performance when using split vs unsplit vertex buffers (avg 708.66fps)

I was able to also capture some traces on my Pixel 4 and while average framerate was identical (avg 13.66fps) I did see a drop in vertex memory bandwidth in a frame (this benefit coming directly from bandwidth improvements during the binning phase I mentioned earlier) where for unsplit I saw peak and average bandwidth of 13.28GB/s and 3.7GB/s respectively, while on the split buffer test I saw peak and average bandwidth of 4.9GB/s and 2.8GB/s respectively.
The improvement seen in the average stems from the improvement during the binning phase of a frame, otherwise during the rendering phase of a frame we see that the vertex bandwidth is pretty much the same (avg 2GB/s) whether split or unsplit (which is to be expected)

I've included links for you to download the Android GPU Inspector trace files I've taken in case you were interested at looking at the data yourself! You just need to have [Android GPU Inspector](https://gpuinspector.dev/) downloaded to take a peek :) 

[vertex buffer split](https://drive.google.com/file/d/1IkvoiPL9DSx_uOoOC6T4Wb4_rJNvmOG2/view?usp=sharing)
[vertex buffer unsplit](https://drive.google.com/file/d/1lqo-IAyd3M8f_IKhnmwxYi7nBhwwbF68/view?usp=sharing)

lawnjelly commented:
> I was able to also capture some traces on my Pixel 4 and while average framerate was identical (avg 13.66fps)

That's quite interesting in itself. I deliberately used identical position triangles and scaled small so that hopefully they would end up in the same tile. The idea being that this would stress the tile based renderer the most as it didn't get the opportunity to spread the geometry over several tiles. But perhaps there is something we are missing and we could run a better test - I'm no expert on tiled renderers. I'm sure a hardware guy could tell us more.

The-O-King commented:
Hello! Just wanted to bump this and see what the status is right now? Sorry it's been a while I got pulled into other work that's taken up my time

So what are the current concerns about this change? Are we still unsure about it's benefits? My current understanding is that on mobile we see the memory bandwidth reductions we were expecting (and implicitly that is battery life improvement as well) and on desktop performance was largely the same (at least on the hardware we've tested so far)

The-O-King commented:
@clayjohn hello! Just wanted to bump this again, and see if there was anything I could do to help :) Thanks!

akien-mga commented:
There's a couple new project settings added by this PR which should be included in the docs, see the diff here: https://github.com/godotengine/godot/pull/46574/checks?check_run_id=2559453240

You can generate the template with `godot --doctool` with a build of your PR, and then ideally fill the descriptions.

clayjohn reviewed (approved):
Approved, subject to addition of docs for the project settings.

The-O-King commented:
Update the documentation with doctool so hopefully things are ready after this!

Thanks for taking a look everyone!

akien-mga commented:
Could you rebase to squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) for instructions.

The-O-King commented:
Oh thank you for letting me know about that I hadn't seen that before, was able to squash everything, let me know if everything is now in order

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -1187,6 +1187,9 @@
 			The default compression level for lossless WebP. Higher levels result in smaller files at the cost of compression speed. Decompression speed is mostly unaffected by the compression level. Supported values are 0 to 9. Note that compression levels above 6 are very slow and offer very little savings.
 		</member>
+		<member name="rendering/mesh_storage/split_stream" type="bool" setter="" getter="" default="false">
+			On import, mesh vertex data will be split into two streams within a single vertex buffer, one for position data and the other for interleaved attributes data. Recommended to be enable if targeting mobile devices. Requires manual reimport of meshes after toggling
```
```suggestion
			On import, mesh vertex data will be split into two streams within a single vertex buffer, one for position data and the other for interleaved attributes data. Recommended to be enabled if targeting mobile devices. Requires manual reimport of meshes after toggling.
```

The-O-King replied:
Updated the word here!

akien-mga replied:
You missed it in my diff but there's also a dot that should be added at the end of the sentence.

The-O-King replied:
Oh shoot you're so right sorry about that I hope that takes care of it now!

akien-mga replied:
I think you might have missed pushing that last amend?

The-O-King replied:
Ah you're right I noticed in the logs that I was getting unresolved host errors in my repo for some reason

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
